messages/README.md

```diff
@@ -0,0 +1,190 @@
+basic typechecking is done via a [semantics plugin](compiler/semantics.py).
+
+Code generation is performed by walking the AST and generating strings containining the messages as types in the language being generated, as well as serialization and deserialization code for each message. In order to decouple the implementation of the AST from the code generation, where generation for each language may be written by different developers, a [Visitor](compiler/visitor.py) pattern is used. Each code generator implements the visitor for a given language that allow it to take callbacks about specific types and generate the corresponding code. Currently there is only a single [code generator for C++](compiler/cpp/cppgen.py), although python is coming soon.
+
+The great thing about generating code via a visitor, is that [tests can be generated as well](test_cppgen.py)!
+
+# Usage
+
+Messages are defined in concord message format (.cmf) files. For C++ a single `.cmf` file will generate a standalone `.h` file. The only dependency is a C++17 standard library.
+
+Generate C++ code:
+
+```bash
+./cmfc.py --input ../example.cmf --output example.h --language cpp --namespace concord::messages
+```
+
+Test C++ code generation. The following:
+ 1. Generates serialization code for [example.cmf](example.cmf)
+ 2. Generates instances of the structs from the generated example.h using uniform initialization
+ 3. Generates tests functions that round trip serialize and deserialize the instances
+ 4. Compiles the test code using g++
+ 5. Runs the tests
+
+```bash
+cd compiler/cpp
+./test_cppgen.py
+```
+
+
+# Data Types
+## Primitive Data Types
+
+* bool
+* unsigned integers -  uint8, uint16, uint32, uint64
+* signed integers - int8, int16, int32, int64
+* string - UTF-8 encoded strings
+* bytes -  an arbitrary byte buffer
+
+## Compound Data Types
+
+Compound data types may include primitive types and other compound types. We ensure canonical serialization by ordering
+
+ * kvpair - Keys must be primitive types. Values can be any type
+ * list - A homogeneous list of any type
+ * map - A lexicographically sorted list of key-value pairs
+ * oneof - A sum type (tagged union) containing exactly one of the given messages. oneof types cannot contain primitives or compound types, they can only refer to messages. This is useful for deserializing a set of related messages into a given wrapper type. A oneof maps to a `std::variant` in c++.
+ * optional - An optional value of any type. An optional maps to a `std::optional` in C++.
+
+## Comments
+
+Comments must be on their own line and start with the `#` character. Leading whitespace is allowed.
+
+ # Serialized Representations
+
+* `bool` - `0x00` = False, `0x01` = True
+* `uint8` - The value itself
+* `uint16` - The value itself in little endian
+* `uint32` - The value itself in little endian
+* `uint64` - The value itself in little endian
```

Comment:
Just curious - is the value that travels over the wire is in little endian? If so - why? Network packets are usually encoded in big endian.

Reply:
The value that travels over the wire is indeed in little endian.

It's a somewhat arbitrary choice biased for common architectures. Big endian was chosen a long time ago when most machines were big endian. That is no longer the case. Since we define the protocol, there's no reason we can't pick the opposite :)

The rationale is that all Intel machines and clones are little endian and we can optimize the code to do a memcpy rather than a byte by byte conversion on little endian machines, and arm machines can be configured to work in little endian. It also allows for zero copy deserialization on little endian machines, although our current implementation doesn't do that because it's less ergonomic to use and I doubt serialization is going to be our bottleneck in concord. 

@f-squirrel Any big endian machine will do the write thing. This is just what goes out on the wire.

If this is a strong concern, it's trivial to change.

Reply:
I know that most of the modern processors are little-endian but if we are building a network protocol it might be reasonable to sent data in network byte order.
AFAIK, most of the ARMs are little-endian. And from my experience ppl don't like to change the default endianness on their platforms.

Anyways, I don't think we have plans to run our SW on some exotic architecture so let us leave it as is.

Reply:
FWIW Libra canonical serialization uses little endian as well. https://github.com/libra/libra/tree/6a89e827b95405066dc83eec97eca2cb75bc991d/common/canonical-serialization